### PR TITLE
Add SSH demo

### PR DESCRIPTION
**Pull request by Jakob-Naucke**
Add basic documentation, Dockerfile and sandbox config for a demo pod
with SSH public key authentication.

Documentation on encrypting and running the image is omitted as of now,
for this is a more general topic which will require separate attention.

Signed-off-by: Jakob Naucke <jakob.naucke@ibm.com>

/cc @sameo @fitzthum @magowan

We can also explain the setup in more detail, but I'd like your general opinion first. Feedback from code review over "this doesn't belong here" to "I have a better idea"  is welcome.

### CODE REVIEW

**Comment by fitzthum** [comment]
This looks good. I really like how you can match up the fingerprints. I'm not sure if this is the best place for it. There was some mention of having a separate demo repo or at least a demo directory.

**Review by magowan — COMMENTED** [review]
This is a good building block for demos, thanks for getting this moving.

I think we do need to create a high level demos directory and put ssh-demo in there?

I worry about repo idea that we end up with lots of repos with very little in them, If we get too much content we can always split out into a different repo later?

Regardless of course, as content gets added I expect us to need to restructure over time anyway :-)

File: ssh-demo/README.md
Comment (Jakob-Naucke) [reply ssh-demo/README.md:8]: Yes, let me clear that up. I am trying to sidestep mentioning Skopeo or ocicrypt, as our use of those is very much in motion at the moment (might be continued to be used for the _encryption_ side though?).

**Review by magowan — APPROVED** [review]

File: demos/ssh-demo/README.md
Comment (sameo) [inline demos/ssh-demo/README.md:3]: Can we elaborate a little more here on the merits of such demo?

For example by explaining that the container image will be encrypted and that neither the guest nor the host will hold the encryption keys. That the only way for this container image to be run and for anyone to be able to ssh into it will be by having the guest getting the encryption keys from a remote attestation server.

Although It *may* be obvious to us, it may not be so for anyone going through that README?

Comment (Jakob-Naucke) [reply demos/ssh-demo/README.md:3]: Yes, this deserves some more background info. I just didn't want to write a ton before getting some degree of approval about doing this in the first place, but I guess we have that now :)

Comment (magowan) [reply demos/ssh-demo/README.md:3]: I guess it is semantics if it is truely the attestation server that is used to get the keys v a further key broker service?

I would hesitate to prescribe the key must come from a remote attestation server since with at least one TEE technology (Secure Execution) it is possible to secure the key inside a guest image without needing to retrieve it from the attestation flow.
Indicating that the encryption key needs to be protected from the host I completely agree with but I feel protected from the guest needs to be discussed more. 
Fundamentally containers will run inside the guest and we don't claim to protect what is in the guest from the containers (compared to guest from host) , so if my guest image can hold the key in uncompromised way then I must have built the image for it to have the appropriate key hence attestation flow is not as critical. A guest image without the key won't be able to run the container image.
So:
If I build a dedicated guest image which is itself encrypted and provides protection then putting keys in there is reasonable.
If I build a generic guest image then this would raise concerns on how generic and start to strongly suggest attestation.
And of course if I cannot use encrypted guest images then attestation flow becomes our delivery mechanism.

Comment (fitzthum) [reply demos/ssh-demo/README.md:3]: Well at the very least we should probably try to explain the difference between using exec to access the container and using SSH with this setup. We are creating a secure channel between the guest owner and the workload. This is mediated by a trusted entity. At this level there is really no difference between using a KBC that gets the secret via SEV secret injection and a KBC that gets the secret via an encrypted initrd. Those are two different secure mechanisms for provisioning the secrets that trust a non-csp entity. In general it might be good to focus on how these platforms are fundamentally aligned.

**Comment by Jakob-Naucke** [comment]
Okay, made some updates, also includes a k8s deployment now (unfortunately, I cannot specify an image at this time; I also want to publish an example that could be put there, but I haven't done that yet). I tried to remain technology-agnostic in my added comments; counter-suggestions are welcome. I had also considered an ASCII drawing (or sth like that), DYT this would fit or just introduce confusion?

**Review by sameo — APPROVED** [review]

**Comment by Jakob-Naucke** [comment]
an infinite wait is required for k8s

**Comment by Jakob-Naucke** [comment]
> > an infinite wait is required for k8s
> 
> What do you mean here?

@sameo ah, sorry. I had added a `&& sleep inf` to the entrypoint again to avoid container exit (you often see a `tail -f /dev/null` instead), which is required for e.g. Docker, Podman, Kubernetes, but wasn't required with `crictl`. I figured out that a better way to do this is to not daemonise sshd, which I have pushed now.

**Comment by sameo** [comment]
> > > an infinite wait is required for k8s
> > 
> > 
> > What do you mean here?
> 
> @sameo ah, sorry. I had added a `&& sleep inf` to the entrypoint again to avoid container exit (you often see a `tail -f /dev/null` instead), which is required for e.g. Docker, Podman, Kubernetes, but wasn't required with `crictl`. I figured out that a better way to do this is to not daemonise sshd, which I have pushed now.

I see, thanks for the explanation :-)

**Pull request merged**
